### ISSUE TITLE
Add support for `for` and `if` in `GridLayout`

### ISSUE THREAD

**Issue by tronical** [sixtyfpsui/sixtyfps#358]
After #349 outlined that there were build errors - and those are fixed now - it would still be desirable to use `if` and `for` in grid layouts.

**Comment by FlorentBecker** [comment]
Would it be possible to have a StaticGridLayout with a fixed num_rows and num_cols which accepts for loops?

**Comment by ogoffart** [comment]
> Would it be possible to have a StaticGridLayout with a fixed num_rows and num_cols which accepts for loops?

Yeah, i think this would make sense to have these properties in the GridLayout.
we would also need a flow-direction property then.

**Comment by Tastaturtaste** [comment]
Has there been any progress on this one? What is a possible workaround?

**Comment by ogoffart** [comment]
Right now, the workaround is to use HorizontalLayout in VerticalLayout.  Or to do the layouting "manually" using x, y bindings

**Comment by coufxr** [comment]
I'm sorry, I would like to know if there are other dynamic methods with better performance than using `for` in `Rectangle`? Or, could you tell me what the current best method is for dynamically drawing grids?

**Comment by petrostrak** [comment]
> I'm sorry, I would like to know if there are other dynamic methods with better performance than using `for` in `Rectangle`? Or, could you tell me what the current best method is for dynamically drawing grids?

Would love that too, thanks!

**Comment by mfwre** [comment]
Where there any further developments?

**Comment by ubruhin** [comment]
Same here, I regularly miss this feature.

**Comment by jwtocadero** [comment]
A solution could be to add a `nr-columns` property  to `GridLayout`

```
GridLayout {
    nr-columns: 2
    for item in model: Rectangle {
        // ...
    }
}
```

and only allow `for `if that is set. This could cover many use cases.

**Comment by dfaure-kdab** [comment]
> Would it be possible to have a StaticGridLayout with a fixed num-rows and num-cols which accepts for loops?

Sounds good to me, for [tronical's use cases](https://github.com/slint-ui/slint/issues/358#issuecomment-946490625) number 3 and 4, i.e. to turn each item in the model into a cell in the grid. Use cases 1 and 2 (where an item in the model is represented by a whole row or column in the grid) have a different syntax anyway, so they can be implemented separately.

> we would also need a flow-direction property then

I'm wondering if we do. We could also say: if you set a fixed number of rows then the layout will be vertical (fill the N rows in column 0, first, then the N rows in column 1) and if you set a fixed number of columns then the layout will be horizontal (fill the N columns for row 0 first, then move on to the next row). Pretty much like the previous comment expects too.

**Comment by ogoffart** [comment]
The reason the `col` and `row` and `*span` need to be constant was to simplify the implementation but it doesn't have to be that way.
We should support the following:

```slint
// the col and row comes from the model
GridLayout {
   for foo in items: SomeWidget {
       col: foo.pos-x;
       row: foo.pos-y;
       data: foo.data;
   }
}


// fill Row
GridLayout {
   Row { Widget1{} }
   Row { 
      for foo in items: SomeWidget {
          data: foo.data;
      }
   }
   Row { 
      for foo in items: SomeOtherWidget {
          data: foo.data;
      }
   }

}


// Maybe the same with Columns?
GridLayout {
   Column { 
      for foo in items: SomeWidget {
          data: foo.data;
      }
   }
   Column { 
      for foo in items: SomeOtherWidget {
          data: foo.data;
      }
   }
}


// We can also generate Row from a model
GridLayout {
   for foo in items: Row { 
      SomeWidget {
          data: foo.data;
      }
      SomeOtherWidget { /*...*/ }
   }
}


// Or we'd have automated layouting with a num-column property
GridLayout {
   num-column: 5;
   for foo in items: SomeWidget { data: foo; }
}
```


At runtime, the grid layout algorogthm can already support all of that. All that's missing is the compile time ability to generate the right data structure.
See the `grid_layout_cell_data` function in llr/lower_expression.rs.  In there, col_or_row is a numberliteral, but it could be any expression if GridLayoutElement was holding epxression instead of actual number. 
And the LayoutItem::element could be repeated, but then we'd need to iterate over the the repeater (generate code that iterate over the repeater). We do that for the horizontal/vertical layout and we could do the same for grid.

**Comment by ubruhin** [comment]
>  We should support the following:

That's exactly what I'm looking for, would be awesome if all these variants were supported! 🎉

**Comment by dfaure-kdab** [comment]
`col_or_row` is what makes this much more complex than in the case of horizontal/vertical layouts...                                                                                            

Here's my current thinking:
Before even looking at if/for, we should allow `row` or `col` to be an expression instead of a constant (e.g. setting `col` to the value of a property on another slint object), as @ogoffart suggested
But when I started to implement this, it raised a number of questions, due to the mixture of determining row/col at compile time (col +=1 when nothing else is set, new row when a Row element is used...) and the support for non-constant row/col determined at runtime.

Examples:
```
GridLayout {
  Row {
    Rectangle { background: black; col: some-prop; }
    Rectangle { background: blue; }
  }
}
```
If some-prop is set to 5 at runtime, where is the blue rectangle? In column 0? 1? 6?

Another one:
```
Rectangle { background: red; row: my-row-number; }
Rectangle { background: green; }
```
If my-row-number changes from 3 to 5, what happens to the green rectangle? Does it follow along to that new row (as would happen if it said `row: 5`, or does it stay in the cell which was determined at compile time?

I see three options:
1. the cells with a variable row/col number move around independently from all others, which remain determined at compile time "the old way" (so the answers to the above cases are 0 and 0, not very intuitive I guess - but easy to implement)
2. at compile time we automatically create expressions like `col: some-prop + 1;` and `row: my-row-number + 1;` so that the subsequent cells move along when the property changes (or `col: previous_cell.col + 1`, somehow, to avoid evaluating a complex value many times, if some-prop was a function call instead?)
3. at compile time we only store very basic information (col expression or not set, row expression or not set, first cell in a Row element), and at runtime we go through it all to figure out what goes where (this is however a large departure from the current design of generating two instances of `GridLayoutCellData` (one h, one v) each with `col_or_row` (even if that's changed from constant to expression, it doesn't really lend itself to a runtime step that determines cell positions, I think).

Am I on the right track at all, thinking about col/row expressions as a first step?
In the actual case of for/if support, there won't be such a mix of compile-time vs runtime values, except in corner cases (one can still do that, but none of the actual examples in this issue do that).

**Comment by 0x6e** [comment]
My gut instinct is that (2) matches the behaviour I would expect. I.e. the unspecified row/col properties increment from the previous value if not set, or a new row hasn't been started.

**Comment by ogoffart** [comment]
Thanks @dfaure-kdab to look into this.
I tend to agree that 2. is the behavior we want. Especially if someone does somewhere `Text { text: some-element.col; } ` But the issue is that each of these are costly, so it would be nice if we only had them when required.
I think 3. is a good optimization and has the same behavior from the point of view of the user.
So I'd favor 3, but we would still need code that generate the default `col` or `row` property when it is materialized.

**Comment by dfaure-kdab** [comment]
Right, (2) and (3) lead to the same user-visible behavior (row=0,col=6 and row=5,col=1 in my two examples).
(and indeed, I'm testing the use of those properties in `text` properties)

When you say "when it is materialized", that's at compile time, so that those properties can be used in bindings, right? Which means defining them with a fake value (e.g. 0) in `materialize_fake_properties.rs`, and updating them at runtime in `solve_grid_layout()`, which will trigger bindings re-evaluations? I don't see another solution, just making sure.

**Comment by ogoffart** [comment]
> When you say "when it is materialized", that's at compile time, so that those properties can be used in bindings, right?

What I was thinking is the following:
 - The generated code that calls solve_grid_layout tries to determine the row or col at runtime based on the sibling items if it has a "dynamic" value. We wouldn't set the `col` and `row` property.
 - Only in materialize_fake_property we would assign an expression such as `previous_cell.col + 1` (this is only called if the property is read by another binding).  https://github.com/slint-ui/slint/blob/f58e0d721da143748f681395f9828c337b861fe7/internal/compiler/passes/materialize_fake_properties.rs#L188

But if you want to do it differently that's fine too.

**Comment by dfaure-kdab** [comment]
What if the row number of that cell is changed later, to a new empty row, meaning there is no "previous cell"? The value of the property needs to be updated. `materialize_fake_property` is called only once per property, right?
Isn't it simpler to set the value of the property to a constant that is updated by solve_grid_layout? (which needs to be called again in any case when a row/col property changes, I don't know how that works, but presumably it's already called again if something changes...)

**Comment by dfaure-kdab** [comment]
I managed to move the computation of row and col numbers to runtime rather than compile time in https://github.com/slint-ui/slint/pull/9572
However I am unsure about this whole business of the row and col properties (currently missing from that MR, so I'm very interested in feedback on the MR, even if it can't be merged as is).

Even if we leave dynamic updates for a later stage, row and col properties are already tricky.
My idea would be to store the col and row numbers determined by `determine_row_col_numbers()` into the corresponding Elements (I guess this means adding a ref to the Element in GridLayoutCellData?)... However this might be slower than your idea of doing it "on demand"... but I'm really unsure about how to do that. `materialize_fake_property` is at compile time and doesn't know how to get the values determined at runtime -- if you look at the algorithm in `determine_row_col_numbers()`, it's clearly more complex than `previous_cell.col + 1`, I don't think we want to express (and duplicate) this algorithm using such expressions.

**Comment by ogoffart** [comment]
I'm not expecting that many people would use the `row` and `col` property. So I think we should optimize for the case where these are not needed in properties at runtime. (Or do you think i'm wrong in this assumption?)
I guess one way to do it would be something like pseudo-code `col: layout_cache[cell_index].col`, but then we'd need to store the addition col and row for every cell in the heap, regardless if it is used or not. Perhaps that's not too much so that'd be fine.

> It's clearly more complex than previous_cell.col + 1, I don't think we want to express (and duplicate) this algorithm using such expressions.

I see your point. 
I guess it would be: `col: new_row || previous_cell.row != self.row ? 0 : previous_cell.col + 1` and `row: new_row ? previous_cell.row + 1 : previous_cell.row;` 
More complicated but also not too bad.
You can put it in `materialize_fake_property` or in `insert_fake_property` (the former is lazy and will only construct the binding if needed. The later will always create the binding but it should be removed by later pass if not used.)

But this is just my feeling of the situation. I don't know what's the best. My gut feeling tells me that a few extra binding in very few cases are better than extra heap allocation for every layout. But that said, the bindings also have costs so if `col` and `row` are often used, it would be clearly worse.

**Comment by dfaure-kdab** [comment]
OK but I have no idea how to implement `previous_cell`... And this means that asking for the col number of a cell in column 10 has to traverse (and evaluate properties for) all cells from 10 to 0, kind of reverse iteration in a linked list.

It seems much simpler to me (and acceptable in terms of memory usage) to store two numbers for each cell (e.g. in the layout_cache as you suggested), and materialize on demand a property that looks up into that data structure, doesn't it?

**Comment by ogoffart** [comment]
I see, there is no easy way to access the parent right now from the `initialize` function. The [`find_parent_element`](https://github.com/slint-ui/slint/blob/ae9bf9ef30cc8aeae711733281c50954e6b04a83/internal/compiler/object_tree.rs#L2171C8-L2171C27) function might work but is slow. Might be Ok if there is not too many of these properties.  Maybe it is time to speed up that function by having a weak reference to the parent (but that may be hard as we would need to keep it updated)
So it'd be easier to do it with `insert_fake_property` in the layout.

> It seems much simpler to me (and acceptable in terms of memory usage) to store two numbers for each cell (e.g. in the layout_cache as you suggested), and materialize on demand a property that looks up into that data structure, doesn't it?

Ok, if you think so, then let's go with that.

**Comment by ogoffart** [comment]
I think we can close this issue now that https://github.com/slint-ui/slint/pull/10263 is merged.

**Comment by dfaure-kdab** [comment]
Unfortunately not, support for repeating entire Rows isn't there yet.

**Comment by ogoffart** [comment]
Created https://github.com/slint-ui/slint/issues/10377 for repeating Row.
